primary/src/helper.rs

```diff
@@ -63,35 +72,128 @@ impl<PublicKey: VerifyingKey> Helper<PublicKey> {
                     self.process_certificates(certificate_ids, requestor, true)
                         .await;
                 }
+                // A request that another primary sends us to check whether we have
+                // available to serve the payloads for the provided certificate_ids.
+                PrimaryMessage::PayloadAvailabilityRequest {
+                    certificate_ids,
+                    requestor,
+                } => {
+                    self.process_payload_availability(certificate_ids, requestor)
+                        .await;
+                }
                 _ => {
                     panic!("Received unexpected message!");
                 }
             }
         }
     }
 
+    /// Checks for each provided digest of the list of digests they node has
+    /// the corresponding certificate within its storage and the payload as
+    /// well. If yes, then the requestor can follow up and request the batches
+    /// payload from the primary's workers via the standard syncing approach.
+    async fn process_payload_availability(
+        &mut self,
+        digests: Vec<CertificateDigest>,
+        origin: PublicKey,
+    ) {
+        // get the requestor's address.
+        let address = match self.committee.primary(&origin) {
+            Ok(x) => x.primary_to_primary,
+            Err(e) => {
+                warn!("Primary origin node not found in committee: {e}");
+                return;
+            }
+        };
+
+        let mut result: Vec<(CertificateDigest, bool)> = Vec::new();
+
+        let certificates = match self.certificate_store.read_all(digests.to_owned()).await {
+            Ok(certificates) => certificates,
+            Err(err) => {
+                error!("Error while retrieving certificates: {err}");
+
+                // just return at this point. Send back to the requestor
+                // that we don't have availability - ideally we would like
+                // to communicate an error (so they could potentially retry).
+                result = digests.into_iter().map(|d| (d, false)).collect();
+
+                let message = PrimaryMessage::<PublicKey>::PayloadAvailabilityResponse {
+                    payload_availability: result,
+                    from: self.name.clone(),
+                };
+                self.primary_network
+                    .unreliable_send(address, &message)
+                    .await;
+
+                return;
+            }
+        };
+
+        let certificates_and_ids: Vec<(CertificateDigest, Option<Certificate<PublicKey>>)> =
+            digests.into_iter().zip(certificates).collect();
+
+        for (id, certificate_option) in certificates_and_ids {
```

Comment:
You're materializing the iterator with `collect()` and immediately converting this Vec into another iterator. Consider iterating on the zip instead.